tree.go

```diff
@@ -567,53 +566,81 @@ func (n *InternalNode) Commitment() *Point {
 	return n.commitment
 }
 
+func (n *InternalNode) fillLevels(levels [][]*InternalNode) {
+	levels[int(n.depth)] = append(levels[int(n.depth)], n)
+	for _, child := range n.children {
+		if childInternalNode, ok := child.(*InternalNode); ok && len(childInternalNode.cow) > 0 {
+			childInternalNode.fillLevels(levels)
+		}
+	}
+}
+
 func (n *InternalNode) Commit() *Point {
-	if len(n.cow) != 0 {
-		polyp := frPool.Get().(*[]Fr)
-		poly := *polyp
-		defer func() {
-			for i := 0; i < NodeWidth; i++ {
-				poly[i] = Fr{}
+	if len(n.cow) == 0 {
+		return n.commitment
+	}
+
+	internalNodeLevels := make([][]*InternalNode, StemSize)
+	n.fillLevels(internalNodeLevels)
+
+	points := make([]*Point, 0, 1024)
+	cowIndexes := make([]int, 0, 1024)
+	poly := make([]Fr, NodeWidth)
+	for level := len(internalNodeLevels) - 1; level >= 0; level-- {
+		nodes := internalNodeLevels[level]
+		if len(nodes) == 0 {
+			continue
+		}
+		points = points[:0]
+		cowIndexes = cowIndexes[:0]
+
+		// For each internal node, we collect in `points` all the ones we need to map to a field element.
+		// That is, for each touched children in a node, we collect the old and new commitment to do the diff updating
+		// later.
+		for _, node := range nodes {
+			for idx, nodeChildComm := range node.cow {
+				points = append(points, nodeChildComm)
+				points = append(points, node.children[idx].Commitment())
+				cowIndexes = append(cowIndexes, int(idx))
 			}
-			frPool.Put(polyp)
-		}()
-		emptyChildren := 256
-
-		var i int
-		b := make([]byte, len(n.cow))
-		points := make([]*Point, 2*len(n.cow))
-		for idx, comm := range n.cow {
-			emptyChildren--
-			points[2*i] = comm
-			points[2*i+1] = n.children[idx].Commit()
-			b[i] = idx
-			i++
 		}
```

Comment:
This is step 1. We iterate in all the nodes in the level, and collects in `points` the previous (cow) and new commitments that we'll have to Fr to do the diff updating later.

`cowIndexes` is an auxiliary slice that indicates that the i-th element corresponds to the idx-th element in the node cow. This is needed since doing `range node.cow` has random order (since it's a map). So we can't infer `idx` from the iteration order!